mesh/mesh.go

```diff
@@ -207,29 +211,52 @@ func (m *Mesh) GetLayer(index types.LayerID) (*types.Layer, error) {
 	return l, nil
 }
 
-func (m *Mesh) HandleLateBlock(b *types.Block) {
-	m.Info("Validate late block %s", b.Id())
-	oldPbase, newPbase := m.MeshValidator.HandleLateBlock(b)
-	m.pushLayersToState(oldPbase, newPbase)
-	if err := m.PersistTortoise(); err != nil {
-		m.Error("could not persist Tortoise on late block %s from layer index %d", b.Id(), b.Layer())
-	}
+type validator struct {
+	*Mesh
+	processedLayer types.LayerID
+}
+
+func (m *validator) ProcessedLayer() types.LayerID {
+	defer m.lvMutex.RUnlock()
+	m.lvMutex.RLock()
+	return m.processedLayer
 }
 
-func (m *Mesh) ValidateLayer(lyr *types.Layer) {
-	m.Info("Validate layer %d", lyr.Index())
-	oldPbase, newPbase := m.HandleIncomingLayer(lyr)
+func (m *validator) SetProcessedLayer(lyr types.LayerID) {
+	m.Info("processed layer %d set processed layer %d", m.processedLayer, lyr)
+	defer m.lvMutex.Unlock()
 	m.lvMutex.Lock()
-	m.processedLayer = lyr.Index()
-	if err := m.PersistTortoise(); err != nil {
-		m.Error("could not persist Tortoise layer index %d", lyr.Index())
+	m.processedLayer = lyr
+}
+
+func (v *validator) HandleLateBlock(b *types.Block) {
+	v.Info("Validate late block %s", b.Id())
+	oldPbase, newPbase := v.trtl.HandleLateBlock(b)
+	v.pushLayersToState(oldPbase, newPbase)
+	if err := v.trtl.PersistTortoise(); err != nil {
+		v.Error("could not persist Tortoise on late block %s from layer index %d", b.Id(), b.Layer())
 	}
-	if err := m.general.Put(PROCESSED, lyr.Index().ToBytes()); err != nil {
-		m.Error("could not persist validated layer index %d", lyr.Index())
+}
+
+func (v *validator) ValidateLayer(lyr *types.Layer) {
+	v.Info("Validate layer %d", lyr.Index())
+	oldPbase, newPbase := v.trtl.HandleIncomingLayer(lyr)
+
+	v.SetProcessedLayer(lyr.Index())
+
+	if len(lyr.Blocks()) == 0 {
+		v.Info("skip validation of layer %d with no blocks", lyr.Index())
+		return
```

Comment:
You only get to this flow if you were not synced (and got a zero-block layer when syncing). If you are synced, listening to the gossip, and no blocks arrived you won't end up here. I don't see a reason to have a different flow for both cases.

Reply:
getAndValidate also calls this method and while we are synced we use getAndValidate